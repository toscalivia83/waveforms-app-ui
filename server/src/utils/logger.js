const winston = require('winston')
const consoleTransport = new winston.transports.Console();
const myWinstonOptions = {
    transports: [consoleTransport]
};
const logger = new winston.createLogger(myWinstonOptions);

const logRequest = function(message) {
    logger.info(message);
};

const logError = function(err) {
    logger.error(err)
};
module.exports = {
    logRequest,
    logError
}