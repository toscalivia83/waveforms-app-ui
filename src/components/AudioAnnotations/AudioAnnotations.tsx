import React from "react";
import { AudioFormDetails } from "../../types/app";
import styles from "./AudioAnnotations.module.css";

interface Props {
  annotations: AudioFormDetails;
}

const AudioAnnotations = ({ annotations }: Props): React.ReactElement => {
  return (
    <div className={styles.container}>
      <div>Number of breaths: {annotations.numberOfBreaths}</div>
      <div>{annotations.hasHeartBeats ? "Has heart beats" : "Does not have heart beats"}</div>
      <div>Audio url: {annotations.audioUrl}</div>
      <br />
    </div>
  );
};

export default AudioAnnotations;