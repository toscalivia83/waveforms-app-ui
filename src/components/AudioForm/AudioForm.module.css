.container {
  margin: 15px 0;
  border-top: 1px solid #EAEDF0;
}

.submitButton {
  background-color: #4CAF50;
}