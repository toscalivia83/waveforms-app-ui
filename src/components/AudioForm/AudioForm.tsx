import React, { useState } from "react";
import { AudioFormDetails } from "../../types/app";

interface Props {
  peaksUrl: string;
  onSubmit: (audioFormDetails: AudioFormDetails) => Promise<void>;
}

const AudioForm = ({ peaksUrl, onSubmit }: Props): React.ReactElement => {
  const [breathsNumber, setBreathsNumber] = useState("0");
  const [hasHeartBeats, setHasHeartBeats] = useState(false);

  const reinitializeAudioFormDetails = (): void => {
    setBreathsNumber("0");
    setHasHeartBeats(false);
  };

  return (
    <div>

      <div>
        <label>Number of breaths</label>
        <input
          name="breathsNumber"
          value={breathsNumber}
          onChange={(event: React.ChangeEvent<HTMLInputElement>): void => {
            console.log("breath nb field changed");
            setBreathsNumber(event.target.value);
          }}/>
      </div>
      <div>
        <label>Presence of heart beats</label>
        <input
          type="checkbox"
          name="termsAndConditions"
          checked={Boolean(hasHeartBeats)}
          onChange={(): void => {
            setHasHeartBeats(!hasHeartBeats);
          }}
        />
      </div>
      <input
        type="submit"
        onClick={(event: React.MouseEvent): void => {
          event.preventDefault();
          onSubmit({ breathsNumber, hasHeartBeats, url: peaksUrl });
          reinitializeAudioFormDetails();
        }} value="Submit"
      />
    </div>
  );
};

export default AudioForm;