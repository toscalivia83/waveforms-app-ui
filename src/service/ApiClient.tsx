import { AudioFormDetails, AudioAnnotationResponse } from "../types/app";
import { HttpError, BaseResponse } from "../types/http";

export class ApiClient {
  public readonly host: string;

  public constructor(baseUrl: string) {
    this.host = baseUrl;
  }

  public async postAudioFormDetails(audioForm: AudioFormDetails): Promise<BaseResponse<string>> {
    const requestOptions = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json"
      },
      body: JSON.stringify(audioForm)
    };
    const uri = `${this.host}/audio-annotations`;
    const response = await fetch(uri, requestOptions);

    try  {
      const responseText = await response.text();

      return {
        statusCode: response.status,
        result: responseText
      };
    } catch (e) {
      const error: HttpError = new Error(`request for ${uri} failed`);
      error.notFound = response.status === 404;
      error.status = response.status;
      throw error;
    }
  }

  public async getAudioAnnotations(): Promise<BaseResponse<AudioAnnotationResponse>> {
    const uri = `${this.host}/audio-annotations`;
    const response = await fetch(uri);

    try  {
      const responseJson = await response.json();

      return {
        statusCode: response.status,
        result: responseJson
      };
    } catch (e) {
      const error: HttpError = new Error(`request for ${uri} failed`);
      error.notFound = response.status === 404;
      error.status = response.status;
      throw error;
    }
  };

}

export default new ApiClient(`${window.location.protocol}//${window.location.host}`);