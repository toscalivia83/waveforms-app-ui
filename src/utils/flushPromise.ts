export const flushPromise = async (): Promise<unknown> => {
  return new Promise<unknown>((resolve: (value?: unknown) => void): NodeJS.Immediate => setImmediate(resolve));
};
