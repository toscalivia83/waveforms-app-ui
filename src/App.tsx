import React, { useState } from "react";
import "./App.css";
import { AudioFormDetails } from "./types/app";
import ApiClient from "./service/ApiClient";
import AudioForm from "./components/AudioForm/AudioForm";
import Peaks from "./components/Peaks/Peaks";

const App = (): React.ReactElement => {
  const [recordId, setRecordId] = useState(1);
  // const [lungSoundData, setLungSoundData] = useState({"hey"});
  const peaksUrl = `https://storage.googleapis.com/annotation_tool_feebris/lung_sound_${recordId}.wav`;

  React.useEffect(() => {
    async function getLungSoundData(): Promise<void> {
      if (recordId < 11 ) {
        const lungSoundDataResponse = await ApiClient.getLungSoundData(recordId.toString());
        console.log("lungSoundDataResponse", lungSoundDataResponse);
      }
      // setLungSoundData(lungSoundDataResponse?.result?.lungSoundData || "");
    }

    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    getLungSoundData();
  }, [recordId]);

  return (
    <div className="App">
      This is my Feebris application
      <h1>Audio Waveform numero {recordId}:</h1>
      <Peaks peaksUrl={peaksUrl}/>
      <AudioForm
        peaksUrl={peaksUrl}
        onSubmit={async (audioFormDetails: AudioFormDetails): Promise<void> => {
          await ApiClient.postAudioFormDetails(audioFormDetails);
          setRecordId(recordId + 1);
        }}/>
    </div>
  );
};

export default App;
