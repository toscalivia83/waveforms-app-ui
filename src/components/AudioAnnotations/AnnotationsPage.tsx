import React, { useState } from "react";
import ApiClient from "../../service/ApiClient";
import { AudioAnnotation } from "../../types/app";
import AudioAnnotations from "./AudioAnnotations";

interface Props {
  displayAnnotation: boolean;
  onClick: Function;
}

const AnnotationsPage = ({ displayAnnotation, onClick }: Props): React.ReactElement => {
  const [audioAnnotations, setAudioAnnotations] = useState([] as AudioAnnotation[]);

  React.useEffect(() => {
    async function getAudioAnnotations(): Promise<void> {
      const AudioAnnotationResponse = await ApiClient.getAudioAnnotations();
      setAudioAnnotations(AudioAnnotationResponse.result?.audioAnnotations || []);
    }

    if (displayAnnotation) {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      getAudioAnnotations();
    }
  }, [displayAnnotation]);

  return (
    <>
      <button
        onClick={(): void => onClick()}>
        {!displayAnnotation ? "Display submitted infos" : "Hide submitted infos"}
      </button>
      {
        displayAnnotation && audioAnnotations.length > 0 &&
        audioAnnotations.map((audioAnnotation, index) =>
          <AudioAnnotations
            key={index}
            annotations={audioAnnotation}/>
        )
      }
      {
        displayAnnotation && !audioAnnotations.length &&
        <div>No audio Annotations registered</div>
      }
    </>
  );
};

export default AnnotationsPage;