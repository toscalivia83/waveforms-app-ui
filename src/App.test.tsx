/* eslint-disable @typescript-eslint/unbound-method */
import React from "react";
import { mount, ReactWrapper } from "enzyme";
import ApiClient from "./service/ApiClient";
import App from "./App";
import { flushPromise } from "./utils/flushPromise";
import { act } from "@testing-library/react";
import AudioForm from "./components/AudioForm/AudioForm";
import AudioAnnotations from "./components/AudioAnnotations/AudioAnnotations";
import AnnotationsPage from "./components/AudioAnnotations/AnnotationsPage";

const CALLED_ONCE = 1;

describe("<App /> suite", () => {
  let wrapper: ReactWrapper;

  it("renders correctly", () => {
    const wrapper = mount(<App recordId={1} onRecordIdChange={jest.fn}/>);
    expect(wrapper.find(AudioForm).exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders App correctly and submits audio form", async () => {
    ApiClient.postAudioFormDetails = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: "sent"
    }));

    act((): void => {
      wrapper = mount(<App recordId={1} onRecordIdChange={jest.fn}/>);
    });

    expect(wrapper.find("audio").prop("src")).toBe("https://storage.googleapis.com/annotation_tool_feebris/lung_sound_1.wav");
    expect(wrapper.find(AudioAnnotations).exists()).toBeFalsy();

    wrapper.find("input[name='numberOfBreaths']").simulate("change", { target: { value: "45" } });
    wrapper.find("input[name='hasHeartBeats']").simulate("change", { target: {
      checked: true,
    } });

    wrapper.find("button[type='submit']").simulate("click");

    await act(async (): Promise<void> => {
      await flushPromise();
      wrapper.update();
    });

    expect(ApiClient.postAudioFormDetails).toHaveBeenCalledTimes(CALLED_ONCE);
    expect(wrapper.find("h1").text()).toBe("Audio Waveform numero 2:");
    expect(wrapper.find("audio").prop("src")).toBe("https://storage.googleapis.com/annotation_tool_feebris/lung_sound_2.wav");
  });

  it("renders no audio annotations text correctly", async () => {
    ApiClient.getAudioAnnotations = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: {
        audioAnnotations: []
      }
    }));

    act((): void => {
      wrapper = mount(<App recordId={1} onRecordIdChange={jest.fn}/>);
    });

    const displayAnnotationsButton = wrapper.find(AnnotationsPage).find("button");
    displayAnnotationsButton.simulate("click");

    await act(async (): Promise<void> => {
      await flushPromise();
      wrapper.update();
    });

    expect(wrapper.find(AudioAnnotations)).toHaveLength(0);
    expect(ApiClient.getAudioAnnotations).toHaveBeenCalledTimes(CALLED_ONCE);
  });

  it("renders expanded and collapsed audio annotations correctly", async () => {
    ApiClient.getAudioAnnotations = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: {
        audioAnnotations: [{
          numberOfBreaths: "12",
          hasHeartBeats: false,
          audioUrl: "url1",
          _id: 12,
        }]
      }
    }));
    ApiClient.postAudioFormDetails = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: "sent"
    }));

    act((): void => {
      wrapper = mount(<App recordId={1} onRecordIdChange={jest.fn}/>);
    });

    const displayAnnotationsButton = wrapper.find(AnnotationsPage).find("button");
    displayAnnotationsButton.simulate("click");

    await act(async (): Promise<void> => {
      await flushPromise();
      wrapper.update();
    });

    expect(wrapper.find(AudioAnnotations)).toHaveLength(1);
    expect(ApiClient.getAudioAnnotations).toHaveBeenCalledTimes(CALLED_ONCE);

    wrapper.find("input[name='numberOfBreaths']").simulate("change", { target: { value: "45" } });
    wrapper.find("input[name='hasHeartBeats']").simulate("change", { target: {
      checked: true,
    } });

    wrapper.find("button[type='submit']").simulate("click");

    await act(async (): Promise<void> => {
      await flushPromise();
      wrapper.update();
    });

    expect(ApiClient.postAudioFormDetails).toHaveBeenCalledTimes(CALLED_ONCE);
    expect(wrapper.find("h1").text()).toBe("Audio Waveform numero 2:");
    expect(wrapper.find("audio").prop("src")).toBe("https://storage.googleapis.com/annotation_tool_feebris/lung_sound_2.wav");
    expect(wrapper.find(AudioAnnotations)).toHaveLength(0);
    expect(ApiClient.getAudioAnnotations).toHaveBeenCalledTimes(CALLED_ONCE);
  });
});
