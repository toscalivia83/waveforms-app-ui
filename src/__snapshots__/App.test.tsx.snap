// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> suite renders correctly 1`] = `ReactWrapper {}`;
